modules/homebrew.nix

```diff
@@ -1,219 +1,751 @@
+          Brewfile are left installed.
+
+          When set to <literal>"uninstall"</literal>, <command>nix-darwin</command> invokes
+          <command>brew bundle [install]</command> with the <command>--cleanup</command> flag. This
+          uninstalls all formulae not listed in generate Brewfile, i.e.,
+          <command>brew uninstall</command> is run for those formulae.
+
+          When set to <literal>"zap"</literal>, <command>nix-darwin</command> invokes
+          <command>brew bundle [install]</command> with the <command>--cleanup --zap</command>
+          flags. This uninstalls all formulae not listed in the generated Brewfile, and if the
+          formula is a cask, removes all files associated with that cask. In other words,
+          <command>brew uninstall --zap</command> is run for all those formulae.
+
+          If you plan on exclusively using <command>nix-darwin</command> to manage formulae
+          installed by Homebrew, you probably want to set this option to
+          <literal>"uninstall"</literal> or <literal>"zap"</literal>.
+        '';
+      };
+      update = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to enable Homebrew to auto-update itself and all formulae during
+          <command>nix-darwin</command> system activation. The default is <literal>false</literal>
+          so that repeated invocations of <command>darwin-rebuild switch</command> are idempotent.
+
+          If you prefer this behavior you many also want to consider enabling
+          ${mkDocOptionLink "homebrew.global.noAutoUpdate"}.
+        '';
+      };
+      upgrade = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to enable Homebrew to upgrade outdated formulae and Mac App Store apps during
+          <command>nix-darwin</command> system activation. The default is <literal>false</literal>
+          so that repeated invocations of <command>darwin-rebuild switch</command> are idempotent.
+        '';
+      };
+
+      brewBundleCmd = mkInternalOption { type = types.str; };
+    };
+
+    config = {
+      brewBundleCmd = concatStringsSep " " (
+        optional (!config.update) "HOMEBREW_NO_AUTO_UPDATE=1"
+        ++ [ "brew bundle --file='${brewfileFile}' --no-lock" ]
+        ++ optional (!config.upgrade) "--no-upgrade"
+        ++ optional (config.cleanup == "uninstall" || config.cleanup == "zap") "--cleanup"
+        ++ optional (config.cleanup == "zap") "--zap"
+      );
+    };
+  };
+
+  globalOptions = { config, ... }: {
+    options = {
+      brewfile = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to enable Homebrew to automatically use the Brewfile that this module generates in
+          the Nix store, when you manually invoke <command>brew bundle</command>.
+
+          Implementation note: when enabled, this option sets the
```

Comment:
Yeah, I had flagged this in the `homebrew.global.noLock` description, which used to say:

```
          Whether to disable lockfile generation when you manually invoke
          <command>brew bundle [install]</command>. This is often desirable when
          ${mkDocOptionLink "homebrew.global.brewfile"} is enabled, since
          <command>brew bundle [install]</command> will try to write the lockfile in the Nix store,
          and complain that it can't (though the command will run successfully regardless).
```

Your comment here did bring to my attention that folks might not see this, and end up enabling `homebrew.global.brewfile` without also enabling `homebrew.global.noLock` when they would have wanted to. In fact, it seems unlikely that folks would want to enable the former without also enabling the latter.

As a result, I've changed the description of `homebrew.global.brewfile` to include,
```
          [...]
          Enabling this option will change the default value of
          ${mkDocOptionLink "homebrew.global.noLock"} to <literal>true</literal> since,
          <command>brew bundle [install]</command> will try to write the lockfile in the Nix store,
          and complain that it can't (though the command will run successfully regardless).
          [...]
```

and the description of `homebrew.global.noLock` to,
```
          [...]
          This option will default to <literal>true</literal> if
          ${mkDocOptionLink "homebrew.global.brewfile"} is enabled, since
          <command>brew bundle [install]</command> will try to write the lockfile in the Nix store,
          and complain that it can't (though the command will run successfully regardless).
          [...]
```

and added the following the the configuration of the `homebrew.global` submodule,
```
      # See description of `homebrew.global.{brewfile,noLock}` options for context.
      noLock = mkIf config.brewfile (mkDefault true);
```

along with a note in the changelog about this.

Reply:
This seems sensible to me 👍